### PR TITLE
Division Tags

### PR DESCRIPTION
**Pull request by andyluo03**

### CODE REVIEW

File: tle/util/cache_system2.py [hunk 1]
```diff
@@ -331,11 +339,24 @@ async def _fetch_problemsets(self, contests, *, force_fetch=False):
...
-            _, problemset, _ = await cf.contest.standings(contest_id=contest_id, from_=1,
+            contest, problemset, _ = await cf.contest.standings(contest_id=contest_id, from_=1,
                                                           count=1)
+            contest_name: str = contest.name
+            contest_name = contest_name.lower()
+            divisions = []
+            for i in range (1, 5):
+                if f'div. {i}' in contest_name:
+                    divisions.append(f'div{i}')
+                    divisions.append(f'd{i}')
```
Comment (Denjell) [inline tle/util/cache_system2.py:350]: divX is enough

File: tle/util/cache_system2.py [hunk 2]
```diff
@@ -331,11 +339,24 @@ async def _fetch_problemsets(self, contests, *, force_fetch=False):
...
     async def _fetch_for_contest(self, contest_id):
         try:
-            _, problemset, _ = await cf.contest.standings(contest_id=contest_id, from_=1,
+            contest, problemset, _ = await cf.contest.standings(contest_id=contest_id, from_=1,
                                                           count=1)
+            contest_name: str = contest.name
+            contest_name = contest_name.lower()
+            divisions = []
+            for i in range (1, 5):
+                if f'div. {i}' in contest_name:
```
Comment (Denjell) [inline tle/util/cache_system2.py:348]: maybe we can increase the robustness if we steal some code from ;vc command
`        recommendations = {contest.id for contest in contests if
                           contest.matches(markers)`
where markers is like our tags. so we can just iterate over the 4 divisions and check 
`if contest_name.matches("div1") `
and ig it works

Comment (Denjell) [reply tle/util/cache_system2.py:348]: matches is a member function on the Contest object

File: tle/util/cache_system2.py [hunk 3]
```diff
@@ -331,11 +339,24 @@ async def _fetch_problemsets(self, contests, *, force_fetch=False):
 
     async def _fetch_for_contest(self, contest_id):
         try:
-            _, problemset, _ = await cf.contest.standings(contest_id=contest_id, from_=1,
+            contest, problemset, _ = await cf.contest.standings(contest_id=contest_id, from_=1,
                                                           count=1)
+            contest_name: str = contest.name
+            contest_name = contest_name.lower()
```
Comment (Denjell) [inline tle/util/cache_system2.py:345]: not needed when using matches

File: tle/util/cache_system2.py [hunk 4]
```diff
@@ -239,6 +239,14 @@ def keep(problem):
         self.problem_by_name = problem_by_name
         self.problems_last_cache = time.time()
 
+        for problem in self.problems:
+            problem_contest: cf.Contest = self.cache_master.contest_cache.contest_by_id.get(problem.contestId)
+            contest_name = problem_contest.name.lower()
+
+            for i in range(1, 5):
```
Comment (Denjell) [inline tle/util/cache_system2.py:246]: also use matches here, ig

**Review by Denjell — CHANGES_REQUESTED** [review]

File: tle/util/cache_system2.py (hunk 1 above)
Comment (andyluo03) [reply tle/util/cache_system2.py:350]: done

File: tle/util/cache_system2.py (hunk 4 above)
Comment (andyluo03) [reply tle/util/cache_system2.py:246]: done

File: tle/util/cache_system2.py (hunk 3 above)
Comment (andyluo03) [reply tle/util/cache_system2.py:345]: done

File: tle/util/cache_system2.py [hunk 5]
```diff
@@ -239,6 +239,13 @@ def keep(problem):
         self.problem_by_name = problem_by_name
         self.problems_last_cache = time.time()
 
+        for problem in self.problems:
+            problem_contest: cf.Contest = self.cache_master.contest_cache.contest_by_id.get(problem.contestId)
+
+            for i in range(1, 5):
```
Comment (Denjell) [inline tle/util/cache_system2.py:245]: The code in both functions is doing the exact same, why is it so different :-). 
I also look for a more pythonic way to do it (for loops are for C++ devs :D)
Something like:
`divisions = [div_tag for div_tag in _DIV_TAGS if contest.matches(div_tag)]
 problem.tags.append(divisions)
`
And _DIV_TAGS can be declared globally for both functions.
` _DIV_TAGS = ['div1', 'div2', 'div3', 'div4']`

File: tle/util/cache_system2.py [hunk 6]
```diff
@@ -239,6 +239,13 @@ def keep(problem):
         self.problem_by_name = problem_by_name
         self.problems_last_cache = time.time()
 
+        for problem in self.problems:
+            problem_contest: cf.Contest = self.cache_master.contest_cache.contest_by_id.get(problem.contestId)
```
Comment (Denjell) [inline tle/util/cache_system2.py:243]: I see why u use type hint but it does not really match the coding style used everywhere else. i would not start using it at single places.

**Review by Denjell — CHANGES_REQUESTED** [review]

**Pull request merged**